03_gamedesign/index.tex

```diff
@@ -1 +1,139 @@
-\chapter{Game Design}
\ No newline at end of file
+\chapter{Game Design}
+
+\section{Implementation}
+\label{sec:GD:implementation}
+
+Implementation design was important as majority the class were involved in writing simulation code. As such, dedicated \emph{infrastructure} engineers from each team were elected to form an infrastructure team responsible for building the central part of the game, otherwise known as the \emph{server}. Each team's infrastructure engineer was tasked to own the implementation of a slice of the server.
+
+One such sub-team was the \textbf{core} infrastructure team, responsible for building the core parts of the game server. Inputs were taken from the entire class to choose the best implementation strategies and design--a solid and simple core foundation was paramount to allow clean continuous integration of code and ideas from all contributors. The subsections below detail implementation specifics of the game's simulation.
+
+\subsection{Architecture}
+\label{sec:GD:implementation:arch}
+
+The structure of the game was closely modelled after a \emph{client-server} model\footnote{\url{https://en.wikipedia.org/wiki/Client-server_model}}, but note ``closely''--whilst the nomenclature was taken directly from the aforementioned model, there are some differences. \emph{Server} and \emph{client} in this context bear the following definitions:
+
+\begin{definition} \label{def:server}
+    The \textbf{server} is the central game runner, responsible for initiating game events (such as a disaster or the start of a turn). Certain events require the actions of agents, in which the server will invoke a function on the agents to receive a response. Further, the server acts as a source-of-truth for the game's state.
+\end{definition}
+
+
+\begin{definition} \label{def:client}
+    Each \textbf{client} is implemented by an agent. The client provides an interface of functions in which the server can invoke. Moreover, clients may also invoke certain functions from the server's interface to read specific game information.
+\end{definition}
+
+The major difference of this architecture to a traditional client-server model is that, in the former, the stateful central server drives the other clients (by triggering events and eliciting responses), as opposed to clients sending stateless requests to a server to receive a response in the latter.

```

Comment:
i talked about IPC later on, so this might not be worthy of mention now?

Reply:
I think its worthwhile to mention because it's a something that someone might easily assume when you say that you have a client server architecture, so i think pointing out now that this (them being in different processes) is NOT the case makes it clearer in the natural order of reading even though it is covered in more detail later

Reply:
sgtm